cloud/cloud_env_impl.h

```diff
@@ -266,6 +273,28 @@ class CloudEnvImpl : public CloudEnv {
   uint64_t FileCacheGetCharge();
   uint64_t FileCacheGetNumItems();
 
+  std::string CloudManifestFile(const std::string& dbname);
+
+  // Roll the cloud manifest happens when follower tries to become the leader.
+  // It will:
+  //
+  // - Generate new delta based on the specified next_file_num, which represents
+  // the snapshot we cut for the new cloud manifest. SST files with filenum >=
+  // next_file_num belongs to the new epoch
+  // - Apply the delta locally
+  // - Upload CLOUDMANIFEST and MANIFEST to s3
+  //
+  // `cloud_manifest_delta` contains the generated cloud manifest delta during rolling.
+  //
+  // NOTE: If any step after(include itself) this, and before we write
+  // `kNewEpoch` with the delta, fails, we would have to reopen the db. Reason is the local cloudmanifest
+  // contains dirty delta generated during rolling and we have to reopen db to clean it up.
+  Status RollCloudManifest(std::string local_dbname, std::string new_cookie,
+                           uint64_t next_file_num,
+                           std::string* cloud_manifest_delta);
+  // Apply the serialized cloud manifest delta locally.
+  Status ApplyCloudManifestDelta(std::string cloud_manifest_delta);
```

Comment:
These two functions are not really orthogonal. How about this:

```
Status ApplyLocalCloudManifestDelta(const std::string& dbname, uint64_t next_file_num, std::string new_epoch);
Status UploadLocalCloudManifest(const std::string& dbname, std::string& new_cookie);
```

This way you run `ApplyLocalCloudManifestDelta()` in two situations -- applying remote cloud manifest delta and generating the cloud manifest delta yourself.

Leader runs `UploadLocalCloudManifest()` to upload its local CM and M to the cloud.

Feel free to let the upper layers generate the new cloud manifest epoch. We can do that in the leaf codebase.

Same with serializing the cloud manifest delta, we can do that in the leaf codebase.

BTW currently our local MANIFEST filename is `MANIFEST-epoch`. I wonder if this should be just `MANIFEST` (note that cloud filename would still be `MANIFEST-epoch`). That way you don't need to rename the manifest file when you roll the new epoch in cloud manifest. Renaming the manifest file is tricky on a running database.

Reply:
why do you think it's tricky? While renaming the file here, there shouldn't be running threads accessing it right?

Reply:
Yeah, but if we can easily avoid having that constraint we should.

Reply:
If we don't have the constraint, wouldn't we have similar issue even if writing to the same manifest file?

There could be two threads updating manifest file at the same time and we would need lock to coordinate them.